Fixed rare malformed chunk encoding caused by filled up conn->sendBuff

errno opened the pull request:
This change should fix the bug with malformed chunks in HTTPD responses.
See #90 for more details.

errno committed:
Fixed rare malformed chunk encoding caused by filled up conn->sendBuff

chmorgan commented on include/libesphttpd/httpd.h:
```diff
@@ -32,9 +32,13 @@ extern "C" {
+#endif
+
+#ifndef HTTPD_SENDBUFF_MAX_FILL
+#define HTTPD_SENDBUFF_MAX_FILL	HTTPD_SENDBUFF_SIZE - 2
```
can you add a comment as to why the -2 is here? I'm pretty sure I know why but I'm thinking about the next person looking at these #defines

errno replied:
Sure. Added in ec01c41.

errno committed:
Added explanatory comment for HTTPD_SENDBUFF_MAX_FILL

chmorgan commented:
@jkent @phatpaul are you guys actively using the library and have the ability to review this PR? code wise it looks ok

phatpaul commented:
Yes, I'll merge it into my project and try to break it.

phatpaul commented:
I got build error:
```
C:/msys32/home/pabbott/lumitec-dev-fw-plibox-esp32/esp-eth-libesphttpd/main/cgi_common.c: In function 'cgiJsonResponseCommonMulti':
C:/msys32/home/pabbott/lumitec-dev-fw-plibox-esp32/esp-eth-libesphttpd/main/cgi_common.c:95:26: error: 'HTTPD_MAX_SENDBUFF_LEN' undeclared (first use in this function)
   size_t max_send_size = HTTPD_MAX_SENDBUFF_LEN / 2; // div/2 to keep some room for headers
                          ^
C:/msys32/home/pabbott/lumitec-dev-fw-plibox-esp32/esp-eth-libesphttpd/main/cgi_common.c:95:26: note: each undeclared identifier is reported only once for each function it appears in
make[1]: *** [/home/pabbott/lumitec-dev-fw-plibox-esp32/Tools/esp-idf/make/component_wrapper.mk:290: cgi_common.o] Error 1
```
Looks like I was using that macro in my own cgi function.  Can we keep that macro available to avoid breaking peoples code?  
I think anyone using that macro probably wants the -2 bytes value.
I propose:
1. Add additional define for convenience: `#define HTTPD_MAX_SENDBUFF_LEN     HTTPD_SENDBUFF_MAX_FILL`
2. Or change name globally of HTTPD_SENDBUFF_MAX_FILL to HTTPD_MAX_SENDBUFF_LEN

I'll do # 1 and continue testing...

phatpaul commented:
Another thing.  Please enclose the defined value in parenthesis.  I found that it broke my (not so great) code:
`size_t max_send_size = HTTPD_MAX_SENDBUFF_LEN / 2;`
expands to
`size_t max_send_size = HTTPD_SENDBUFF_SIZE - 2 / 2; `
which is 2047 and not the 1023 that I expected.

errno committed:
Added HTTPD_MAX_SENDBUFF_LEN foor backward compatibility. Added parenthesis for code safety.

errno commented:
@phatpaul Thank you for testing and valuable input.

Good point about parenthesis. Fixed.

I renamed HTTPD_MAX_SENDBUFF_LEN for the sake of clarity. I thought the old name might be ambigous. I've checked libesphttpd for usages of this macro, but didn't think of other code that might use it.
I've added it back but defined via HTTPD_SENDBUFF_MAX_FILL.

jkent commented:
This looks good to me.  My only complaint is the use of strcpy instead of memcpy.  The libesphttpd source uses strcpy everywhere though, writing nulls past the end of the buffers in rare cases.  I'm planning on addressing this issue in my rework of libesphttpd.
